File: src/gameStates/GameState.java
package gameStates;

import gameObjects.GameObject;

import java.awt.*;
import java.util.LinkedList;

public abstract class GameState {
    protected StateHandler handler;
    public LinkedList<GameObject> objects;

    public GameState(StateHandler handler) {
        this.handler = handler;
        objects = new LinkedList<GameObject>();
        init();
    }
    protected abstract void init();
    protected abstract void tick();
    protected abstract void render(Graphics g);
    protected abstract void keyPressed(int k);
    protected abstract void keyReleased(int k);
    protected void addObject(GameObject obj) {
        objects.add(obj);
    }
    protected void removeObject(GameObject obj) {
        objects.remove(obj);
    }
}

File: public/gameStates/GameState.js
'use strict'

class GameState {
    handler
    gameObjects

    constructor(handler) {
        this.handler = handler
        this.gameObjects = new Array()
    }

    tick(delta) {
        throw new Error('tick() must be overridden in subclass')
    }
    render(ctx) {
        throw new Error('render() must be overridden in subclass')
    }
    keyPressed(keyCode) {
        throw new Error('keyPressed() must be overridden in subclass')
    }
    keyReleased(keyCode) {
        throw new Error('keyReleased() must be overridden in subclass')
    }
}
